src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/TypeMapLazyDictionary.cs

```diff
@@ -0,0 +1,448 @@
+            CallbackContext* context);
+
+        private static TypeName NewTypeNameWithAssembly(TypeName typeName, ReadOnlySpan<char> assemblyName)
+        {
+            AssemblyNameInfo asmNameInfo = AssemblyNameInfo.Parse(assemblyName);
+            TypeName parsedType = typeName.WithAssemblyName(asmNameInfo);
+            Debug.Assert(parsedType.AssemblyName != null);
+            return parsedType;
+        }
+
+        public ref struct Utf16SharedBuffer
+        {
+            private char[]? _backingArray;
+            public Utf16SharedBuffer()
+            {
+                _backingArray = null;
+                Buffer = default;
+            }
+
+            public Utf16SharedBuffer(char[] backingBuffer, int validLength)
+            {
+                _backingArray = backingBuffer;
+                Buffer = new ReadOnlySpan<char>(backingBuffer, 0, validLength);
+            }
+
+            public ReadOnlySpan<char> Buffer { get; init; }
+
+            public void Dispose()
+            {
+                if (_backingArray != null)
+                {
+                    ArrayPool<char>.Shared.Return(_backingArray);
+                }
+            }
+        }
+
+        private static void ConvertUtf8ToUtf16(ReadOnlySpan<byte> utf8TypeName, out Utf16SharedBuffer utf16Buffer)
+        {
+            const int DefaultCharArrayLen = 1024;
+            const int MaxUtf8BytesPerChar = 3;
+            int needed = utf8TypeName.Length > (DefaultCharArrayLen * MaxUtf8BytesPerChar)
+                ? Encoding.UTF8.GetCharCount(utf8TypeName)
+                : DefaultCharArrayLen;
+
+            char[] buffer = ArrayPool<char>.Shared.Rent(needed);
+            int converted = Encoding.UTF8.GetChars(utf8TypeName, buffer);
+            utf16Buffer = new Utf16SharedBuffer(buffer, converted);
+        }
+
+        [UnmanagedCallersOnly]
+        private static unsafe void NewExternalTypeEntry(CallbackContext* context, ProcessAttributesCallbackArg* arg)
+        {
+            Debug.Assert(context != null);
+            Debug.Assert(arg != null);
+            Debug.Assert(arg->Utf8String1 != null);
+            Debug.Assert(arg->Utf8String2 != null);
+
+            try
+            {
+                string externalTypeName = new((sbyte*)arg->Utf8String1, 0, arg->StringLen1, Encoding.UTF8);
+                TypeNameValue targetTypeName = new()
+                {
+                    Utf8TypeName = arg->Utf8String2,
+                    Utf8TypeNameLen = arg->StringLen2,
+                    Utf8AssemblyNameFallback = arg->Utf8String3,
+                    Utf8AssemblyNameFallbackLen = arg->StringLen3
+                };
+                context->ExternalTypeMap.Add(externalTypeName, targetTypeName);
+            }
+            catch (Exception ex)
+            {
+                context->CreationException = ExceptionDispatchInfo.Capture(ex);
```

Comment:
Return false to stop the further processing?

Reply:
Sure.